### PR TITLE
Statistics link between DDS and RTPS layers [11273]

### PR DESCRIPTION
**Pull request by MiguelCompany**
Built on top of #1922

### CODE REVIEW

File: src/cpp/statistics/fastdds/domain/DomainParticipantImpl.hpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+     * Auxiliary method to delete a topic and deregister a type after checking that they are consistent.
+     * @param topic_name string with the topic name.
+     * @param type_name string with the type name.
+     * @return false if the type and topic do not match or if delete_topic fails. Otherwise true.
+     */
+    bool delete_topic_and_type(
+            const std::string& topic_name) noexcept;
+
+    efd::Publisher* builtin_publisher_;
+    std::shared_ptr<DomainParticipantStatisticsListener> statistics_listener_;
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/statistics/fastdds/domain/DomainParticipantImpl.hpp:176]: When building with `-DFASTDDS_STATISTICS=OFF` this line generates an error:

```
error: ‘DomainParticipantStatisticsListener’ was not declared in this scope
  176 |     std::shared_ptr<DomainParticipantStatisticsListener> statistics_listener_;
```

`DomainParticipantStatisticsListener` is only defined if `FASTDDS_STATISTICS` is set.

File: src/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -249,6 +250,20 @@ if (FASTDDS_STATISTICS)
...
+    set(statistics_sources
+        statistics/fastdds/domain/DomainParticipantStatisticsListener.cpp
+        )
+
+    list(APPEND ${PROJECT_NAME}_source_files
+        ${statistics_sources}
+        )
+
+endif()
+
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/CMakeLists.txt]: Fixed on last rebase

File: test/blackbox/common/DDSBlackboxTestsStatistics.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
+#include <fastdds/dds/subscriber/DataReader.hpp>
+#include <fastdds/dds/subscriber/Subscriber.hpp>
+
+#include <fastdds/statistics/topic_names.hpp>
+#include <fastdds/statistics/dds/domain/DomainParticipant.hpp>
+#include <fastdds/statistics/dds/publisher/qos/DataWriterQos.hpp>
+#include <fastdds/statistics/dds/subscriber/qos/DataReaderQos.hpp>
+
+#include <fastrtps/xmlparser/XMLProfileManager.h>
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/blackbox/common/DDSBlackboxTestsStatistics.cpp:31]: It was before I pushed. I forgot to remove this line.

File: src/cpp/statistics/fastdds/domain/DomainParticipantImpl.hpp (hunk 1 above)
Comment (MiguelCompany) [reply src/cpp/statistics/fastdds/domain/DomainParticipantImpl.hpp:176]: That means that almost the whole file contents could be wrapped on a `#ifdef FASTDDS_STATISTICS`

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Partial review. Mostly include header files that are being used.

File: test/mock/rtps/RTPSReader/fastdds/rtps/reader/RTPSReader.h [hunk 4]
```diff
@@ -80,6 +80,24 @@ class RTPSReader : public Endpoint
...
+    {
+        return true;
+    }
+
+    template<typename T>
+    bool remove_statistics_listener(
+            T listener)
+    {
+        return true;
+    }
```
Comment (JLBuenoLopez-eProsima) [inline test/mock/rtps/RTPSReader/fastdds/rtps/reader/RTPSReader.h:97]: These methods issue an unused parameter `listener` warning.

File: src/cpp/fastdds/publisher/PublisherImpl.hpp [hunk 5]
```diff
@@ -96,6 +96,10 @@ class PublisherImpl
     ReturnCode_t set_listener(
             PublisherListener* listener);
 
+    DataWriter* PublisherImpl::create_datawriter(
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/fastdds/publisher/PublisherImpl.hpp:99]:
```suggestion
    DataWriter* create_datawriter(
```

File: test/blackbox/common/DDSBlackboxTestsStatistics.cpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            reader->return_loan(data_seq, info_seq);
+        }
+    } while (total_samples < num_samples);
+
+    std::cout << "Received " << total_samples << " samples on " << topic_name << std::endl;
+}
+
+#endif // FASTDDS_STATISTICS
+
+TEST(DDSStatistics, simple_statistics_datareaders)
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/blackbox/common/DDSBlackboxTestsStatistics.cpp:103]: We can add them in the future. The idea here is to test one of each kind (writer, reader, participant)

File: test/mock/rtps/RTPSReader/fastdds/rtps/reader/RTPSReader.h (hunk 4 above)
Comment (MiguelCompany) [reply test/mock/rtps/RTPSReader/fastdds/rtps/reader/RTPSReader.h]: Fixed on a83100767bd1a442912b89cc371665961276a5dd

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM

**Pull request merged**